docs/architecture/adr-011-light-client-crates-extraction.md

```diff
@@ -0,0 +1,375 @@
+an encoding to `Any` ->
+
+```rust
+use ibc_proto::google::protobuf::Any;
+
+pub trait Header {
+    /// Encode to canonical protobuf representation
+    fn encode_any(&self) -> Any;
+
+    /* ... */
+}
+```
+
+Thankfully, the core modules code doesn't use the `serde` derivations except in logs and errors. Host and light-client
+implementations can optionally choose to downcast to the concrete type and use it's `serde` derivations directly (if
+available).
+
+#### Light client traits cannot have constructors
+
+This restriction comes from the fact that trait methods of object safe traits cannot return `Self`.  
+However, we would need a constructor to be able to create a `ClientState` and `ConsensusState` in the `create_client`
+handler. This can be done using a `where Self: Sized` clause on the trait method.
+
+```rust
+use ibc_proto::google::protobuf::Any;
+
+pub trait ClientState {
+    fn decode(any: Any) -> Box<dyn ClientState> where Self: Sized;
+
+    /* ... */
+}
+```
+
+### Downcasting support
+
+We need the ability to upcast `&dyn T` to `&dyn core::any::Any` to be able to downcast to a concrete type. This can be
+done using a trait with a blanket implementation ->
+
+```rust
+pub mod dynamic_typing {
+    use core::any::Any;
+
+    pub trait AsAny: Any {
+        fn as_any(&self) -> &dyn Any;
+    }
+
+    impl<T: Any> AsAny for T {
+        fn as_any(&self) -> &dyn Any {
+            self
+        }
+    }
+}
+```
+
+All light client traits could then add `AsAny` as a supertrait.
+
+```rust
+use crate::dynamic_typing::AsAny;
+
+pub trait Header: AsAny { /* ... */ }
+```
+
+And usage would look like ->
+
+```rust
+fn downcast_header(h: &dyn Header) -> Result<&TmHeader, Ics02Error> {
+    h.as_any()
+        .downcast_ref::<TmHeader>()
+        .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))
+}
+```
+
+### Special case: removing `AnyClient`
+
+The `AnyClient` enum is special as it is mostly stateless and always created on-the-fly (using `ClientType`) during
+verification in the handlers. e.g. `update_client::process()` ->
+
+```rust
+pub fn process(
+    ctx: &dyn ClientReader,
+    /* ... */
+) -> HandlerResult<ClientResult, Error> {
+    /* ... */
+
+    let client_type = ctx.client_type(&client_id)?;
+
+    let client_def = AnyClient::from_client_type(client_type);
+
+    let client_state = ctx.client_state(&client_id)?;
+
+    /* ... */
+}
+```
+
+This is problematic because it means the module code must be aware of a `ClientType` -> `ClientDef` mapping. This can be
+solved by requiring the `ClientState` trait to provide us with its `ClientDef` implementation.
+
+```rust
+pub trait ClientState {
+    /// Return the associated `ClientDef` implementation
+    fn client_def(&self) -> Box<dyn ClientDef>;
+
+    /* ... */
+}
+```
+
+Now, we can use the `ClientState` instance to get the `ClientDef` implementation ->
```

Comment:
Can we also use this ADR to rename `ClientDef` to, say, `Client` or `LightClient`?

Reply:
Great idea! Interestingly, the spec calls the verification methods on the `ClientType` itself -> https://github.com/cosmos/ibc/tree/master/spec/core/ics-002-client-semantics#create
I wonder if it'd be better to rename `ClientDef` to `ClientType`.

Reply:
As discussed, ibc-go uses `ClientState` as a combination of our `ClientState` and `ClientDef`: it has the state, and the verify functions. I believe we should follow this approach as well.

So basically, remove `ClientDef`, and put all those methods in our new `ClientState` trait.